pkg/plugin/v3/init.go

```diff
@@ -136,6 +151,16 @@ func (p *initPlugin) Validate() error {
 }
 
 func (p *initPlugin) GetScaffolder() (scaffold.Scaffolder, error) {
+	// Add core resource versions to the config if they are not the defaults so we don't unnecessarily
+	// clutter the resulting PROJECT file.
+	if p.crdVersion != defaultCRDVersion {
+		p.config.AddCoreResourceVersions(config.CRDResourceName, []string{p.crdVersion})
```

Comment:
the CRDversion cannot be an attribute of the Resouce?

Reply:
We'd have to ask for CRD/webhook version every `create <api|webhook>` call, which is too verbose imo. Also having a separate `webhooks` field is not scalable; if we added a new type that users could `kubebuildercreate`, would we add another config field?

The `coreResourceVersion` approach is scalable and succinct because it is optional and only one line per type is required.

Reply:
Hi @estroz, 

Thank you for your attention and sorry, shows that I could not explain properly my suggestion :-(. I will try to that less verbose as possible. 

### Suggestion/Possible solution

IMO It should be done in the create plugins. The init only initialized the project and has not the responsibility to scaffold APIs. We are just letting users define the `apiVersion` of a CRD or Webhook as the same way that they can describe the group/kind/version. Then, it is just a new attribute/arg for the object. 

**UX** (view layer)
  
> $kb create api --group ship --version v1beta1 --kind Frigate --api-version= v1beta1
> $kb create webhook --group ship --version v1beta1 --kind Frigate --conversion --api-version=v1beta1

**Model**

```
resources:
- group: cache
  kind: Memcached
  version: v1alpha1
  apiVersion: v1 

webhooks:
- group: cache
  kind: Memcached
  version: v1alpha1
  apiVersion: v1 
``` 

**Note** (business layer)

E.g: 

```go
// to built themplates
if crd.apiversion == v1 { // scaffold X}
if webhook.apiversion == v1 { // scaffold Y}
```

Roles like if the tool allows or not a value or a combination should NOT drive the design of the model. The model ought to be a realistic and meaningful representation. 
  
By respecting the model if we need to change the roles (business layer), we do not introduce breaking changes. 

I mean, if it will or not allow CRD and Weebhooks with diff apiVersion should NOT be defined in the model. Then, if we after decide to allow CRD/Webhook per kind with v2 and v1 in the same project, all will be fine, and users do not need to change/migrate their projects. 

### Answers for the questions raised. 

> We'd have to ask for CRD/webhook version every create <api|webhook> call

Only when the user does not want the default value. 

> Also having a separate webhooks field is not scalable;

In the f2f we spoke about to use the PROJECT file to be able to re-scaffold all files of the project. How can we do that if we have not []webhooks in the PROJECT file? 

IHMO it needs to be stored since it represents the APIs configured for the project. (same reasons of resources) why store the group/kind/version for Resources is scalable and the same is not valid for webhooks? what is the diff between them? 

>  if we added a new type that users could kubebuilder create, would we add another config field?

I agree with you that if we had any chance of having like 50 API's to scaffold via `kubebuilder create` the config/project design is not viable at all and Kubebuilder would need to change and probably not only at this point. However, would be fine we have more 3 or 5. Do not you agree? I cannot see any other api to be added via `kb create` currently, and in the long term based on what the tool does I cannot see like 50 APIs/create options either for it.

> The coreResourceVersion approach is scalable and succinct because it is optional and only one line per type is required.

Now, in this PR and the specific v1/betav1, we thought in NOT allow projects to have the scenario:  CRD A, B with version v1 and C with version v2. However,  is not very possible that tomorrow we might address the need to allow this need? Could not after we merge and to do the release a user describes a scenario that shows for us that it is required to be achieved? we will not need to it 4ver as well. Will, not the k8s api deprecated v1 and provide v2 at some point?
 
**How the design solution here `coreResourceVersion` can address the need `CRD A, B with version v1 and C with version v2` without breaking changes?**  

See:
- Resources represents the resources 
- Webhooks represents webhooks

`coreResourceVersion` is a temporary/helper structure that actually has none representation in this domain. So, because of this, I am afraid that I do not like too much of this idea and we can't just remove it soon and because of this, we need to think in a solution that will have a long life and not only to solve the actual problem. I mean, we need to avoid breaking changes in the future. 

And then, by driving the solution by the domain (DDD principle), we do need to add the complexities to manage, we keep the code implementation simple and it is +maintainable, readable and re-usable.

Reply:
I think I have two main questions here:

The first is: what's the usecase for wanting separate versions per object (I've got some ideas, but I'm curious to hear what folks are thinking)

The second is: could we just do both?  Make `coreResourceVersions` --> `defaultResourceVersions`, and then individual overrides on the `crds` field or whatever could manually override that.  That way, folks can have consistency across a project, but if they need to do manually override (e.g. I'm only using this hook on older or newer clusters), they can?

Reply:
FWIW, I like the general idea of storing the resources that are scaffolded out in the PROJECT file (including webhooks) too :-)

Reply:
1. Do you mean: what's the usecase for having multiple API versions of the same object definition, ex. a `Frigate` type defined as both a `v1` and `v1beta1` CRD in the sample project? If so, then I would guess one is to support old (< v1.16) and new cluster versions.
1. I think the main problem with having different API versions for various objects is managing what kustomize files get scaffolded, and where. The complexity is a real headache to deal with (I gave it a go and gave up).

Overall I think its better to force users towards newer API versions when possible. If they need to have both available then they have a lot of tools to do so (`controller-gen`, scripting, Makefile mods, etc.) Perhaps supporting only one API version at a time is a good place to start; we can make modifications to the future v4 config to support multiple API versions per object definition?

Reply:
Hi  @DirectXMan12,

Really thank you for your inputs and to raise these questions which also might help we clarifies the thoughts. 
 
>  what's the usecase for wanting separate versions per object

- I am as an operator dev would like to create my new APIs with the upper version but I would NOT like to change what is implemented already so that, I can just deprecate the legacy API versions  
- I am as an operator dev would like to make a project that will identify the cluster version and apply the APIs accordingly, so that, my project can still be supportable on the old and new k8s clusters versions
- I am a consumer of the kubebuilder plugins would like that a plugin has only flags whcih are related to its domain, so that I can decide what plugins I'd like to consume or not. Example: I would like to consume the init and create api one, however, the webhook plugins are not important for me, so that I would NOT like that the init plugin has flags which are related to webhooks. ( e.g helm/ansible in SDK do not support webhooks ) See: https://github.com/operator-framework/operator-sdk/blob/master/internal/plugins/helm/v1/plugin.go#L38
- I am as KB maintainer would like to NOT need care about the init plugin flags when I am working on in solution for the other plugins. For example, would like to be able to make the webhook no longer support an api-version X as arg when I would like to still support this value in the create api plugin. 

> The second is: could we just do both? Make coreResourceVersions --> defaultResourceVersions, and then individual overrides on the crds field or whatever could manually override that. That way, folks can have consistency across a project, but if they need to do manually override (e.g. I'm only using this hook on older or newer clusters), they can?

IMHO: 

- The default version used by the `create api` and `create webhook` plugins to scaffold the files ought to be defined in the plugin code, so that, we can have a default version for v2+ and another for v3+ and we will not need a migration guide step to describe to users update their PROJECT file when it changes. 
- The design should allow we have a default version for each plugin(`create api` and `create webhook`) so that, we will not face problems to address scenarios such as the k8s api deprecate the v1 for webhooks and but not for CRDs.

Reply:
I saw your comment @estroz:

> I think the main problem with having different API versions for various objects is managing what kustomize files get scaffolded, and where. The complexity is a real headache to deal with (I gave it a go and gave up). 

IHMO we can have a design solution which allows us to do that but we do not need to solve this problem now. We could add a role in the validation func of the plugin such as we have for the multigroup support at the first moment. E.g if `project.resources.v1` and `--api-version == v1betav1` raise an error this project has api(s) with the apiVersion `v1` so you cannot scaffold v1betav1 .

Reply:
I think it is solved. So, I am closing this one.